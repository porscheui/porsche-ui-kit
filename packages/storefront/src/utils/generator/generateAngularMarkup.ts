import type { StoryState } from '@/models/story';
import type {
  ConfiguratorTagNames,
  ElementConfig,
  EventConfig,
  HTMLElementOrComponentProps,
  HTMLTagOrComponent,
} from '@/utils/generator/generator';

const getAngularCode = (
  states: string | undefined,
  eventHandlers: string | undefined,
  code: string | undefined
) => `import { ChangeDetectionStrategy, Component } from '@angular/core';

@Component({
  selector: 'example',
  template: \`
${code}
  \`,
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: false,
})
export class ExampleComponent {${states ? `\n${states}\n` : ''}${eventHandlers ? `\n${eventHandlers}\n` : ''}}`;

export const generateAngularMarkup = (
  configs: (string | ElementConfig<HTMLTagOrComponent> | undefined)[],
  initialState: StoryState<HTMLTagOrComponent>,
  indentLevel = 3
): string => {
  const results = configs.map((config) => createAngularMarkup(config, initialState, indentLevel));
  const markup = results.map(({ markup }) => markup).join('\n\n');
  const states = results
    .flatMap(({ states }) => states)
    .filter((state) => state)
    .join('\n');
  const eventHandlers = results.flatMap(({ eventHandlers }) => eventHandlers).join('\n');

  return getAngularCode(states, eventHandlers, markup);
};

const createAngularMarkup = (
  config: string | ElementConfig<HTMLTagOrComponent> | undefined,
  initialState: StoryState<HTMLTagOrComponent>,
  indentLevel = 0
): { markup: string; states: string[]; eventHandlers: string[] } => {
  const indent = '  '.repeat(indentLevel);

  if (!config) return { markup: '', states: [], eventHandlers: [] };
  if (typeof config === 'string') return { markup: `${indent}${config}`, states: [], eventHandlers: [] };

  const { tag, properties = {}, events = {}, children = [] } = config;

  const eventEntries = Object.entries(events);
  const propertiesString = generateAngularProperties(properties, eventEntries);

  const eventListenersString =
    eventEntries.length > 0 ? ` ${eventEntries.map(([eventName]) => `(${eventName})="${eventName}"`).join(' ')}` : '';

  const childrenResults = children.map((child) => createAngularMarkup(child, initialState, indentLevel + 1));
  const childMarkup = childrenResults.map(({ markup }) => markup).join('\n');
  const childStates = childrenResults.flatMap(({ states }) => states);
  const childEventHandlers = childrenResults.flatMap(({ eventHandlers }) => eventHandlers);

  const markup =
    children.length > 0
      ? `${'  '.repeat(indentLevel)}<${tag}${propertiesString}${eventListenersString}>\n${childMarkup}\n${'  '.repeat(indentLevel)}</${tag}>`
      : `${'  '.repeat(indentLevel)}<${tag}${propertiesString}${eventListenersString} />`;

  const scripts = eventEntries.length > 0 ? generateAngularControlledScript(tag, eventEntries, initialState) : null;
  const states = scripts ? [scripts.states, ...childStates] : childStates;
  const eventHandlers = scripts ? [scripts.eventHandler, ...childEventHandlers] : childEventHandlers;

  return { markup, states, eventHandlers };
};

type AngularScripts = { states: string; eventHandler: string };

export const generateAngularControlledScript = (
  tagName: string,
  eventEntries: [string, EventConfig][],
  initialState: StoryState<HTMLTagOrComponent>
): AngularScripts => {
  const states = eventEntries
    // Only create state if the current element's tagName is the same as the element the state is applied to e.g. don't create state for p-button onClick open flyout
    .filter(([_, { target }]) => tagName === target)
    .map(([_, { prop }]) => `  ${prop} = ${(initialState?.properties as any)?.[prop]};`)
    .join('\n');

  const eventHandler = eventEntries
    .map(([eventName, { prop, value, eventValueKey, eventType, negateValue }]) => {
      if (eventValueKey) {
        return `  ${eventName}(e: ${eventType}) {
    this.${prop} = ${negateValue ? '!' : ''}e.detail.${eventValueKey};
  }`;
      }
      return `  const ${eventName} = () => {
    this.${prop} = ${negateValue ? '!' : ''}${value};
  }`;
    })
    .join('\n');

  return { states, eventHandler };
};

export const generateAngularProperties = (
  properties: HTMLElementOrComponentProps<HTMLTagOrComponent>,
  eventEntries: [string, EventConfig][]
) => {
  return Object.entries(properties)
    .map(([key, value]) => {
      if (eventEntries.some(([_, { prop }]) => prop === key)) {
        return ` [${key}]="${key}"`;
      }
      if (typeof value === 'string') return ` ${key}="${value}"`;
      if (key === 'aria') {
        return ` [${key}]="${JSON.stringify(value).replace(/"/g, "'")}"`;
      }
      return ` [${key}]="${JSON.stringify(value)}"`;
    })
    .join('');
};
