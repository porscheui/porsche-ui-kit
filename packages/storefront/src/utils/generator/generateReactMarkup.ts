import type { StoryState } from '@/models/story';
import type {
  ConfiguratorTagNames,
  ElementConfig,
  EventConfig,
  EventsConfig,
  HTMLElementOrComponentProps,
  HTMLTagOrComponent,
} from '@/utils/generator/generator';
import { pascalCase } from 'change-case';

const getReactCode = (
  imports: string,
  states: string | undefined,
  eventHandlers: string | undefined,
  code: string | undefined
) => `${imports}

export const Example = () => {${states ? `\n${states}\n` : ''}${eventHandlers ? `\n${eventHandlers}\n` : ''}
  return (
    <>
${code}
    </>
  )
}`;

export const generateReactMarkup = (
  configs: (string | ElementConfig<HTMLTagOrComponent> | undefined)[],
  initialState: StoryState<HTMLTagOrComponent>,
  indentLevel = 3
): string => {
  const results = configs.map((config) => createReactMarkup(config, initialState, indentLevel));
  const markup = results.map(({ markup }) => markup).join('\n\n');
  const states = results
    .flatMap(({ states }) => states)
    .filter((state) => state)
    .join('\n');
  const eventHandlers = results.flatMap(({ eventHandlers }) => eventHandlers).join('\n');
  const pdsComponents = new Set(results.flatMap(({ pdsComponents }) => pdsComponents));
  const imports = `${states.length > 0 ? "import { useState } from 'react';\n" : ''}${
    pdsComponents.size > 0
      ? `import { ${Array.from(pdsComponents).sort().join(', ')} } from '@porsche-design-system/components-react';`
      : ''
  }`;

  return getReactCode(imports, states, eventHandlers, markup);
};

const createReactMarkup = (
  config: string | ElementConfig<HTMLTagOrComponent> | undefined,
  initialState: StoryState<HTMLTagOrComponent>,
  indentLevel = 0
): { markup: string; states: string[]; eventHandlers: string[]; pdsComponents: string[] } => {
  if (!config) return { markup: '', states: [], eventHandlers: [], pdsComponents: [] };
  const indent = '  '.repeat(indentLevel);

  if (typeof config === 'string')
    return { markup: `${indent}${config}`, states: [], eventHandlers: [], pdsComponents: [] };

  const { tag, properties = {}, events = {}, children = [] } = config;
  const isPDSComponent = tag.startsWith('p-');

  let transformedTag: string = tag;
  const pdsComponents: string[] = [];
  if (isPDSComponent) {
    transformedTag = pascalCase(tag);
    pdsComponents.push(transformedTag);
  }

  const eventEntries = Object.entries(events);
  const propertiesString = generateReactProperties(properties, eventEntries);

  const eventListenersString =
    eventEntries.length > 0 ? ` ${eventEntries.map(([eventName]) => `${eventName}={${eventName}}`).join(' ')}` : '';

  const childrenResults = children.map((child) => createReactMarkup(child, initialState, indentLevel + 1));
  const childMarkup = childrenResults.map(({ markup }) => markup).join('\n');
  const childStates = childrenResults.flatMap(({ states }) => states);
  const childEventHandlers = childrenResults.flatMap(({ eventHandlers }) => eventHandlers);
  const childPDSComponents = childrenResults.flatMap(({ pdsComponents }) => pdsComponents);

  const markup =
    children.length > 0
      ? `${'  '.repeat(indentLevel)}<${transformedTag}${propertiesString}${eventListenersString}>\n${childMarkup}\n${'  '.repeat(indentLevel)}</${transformedTag}>`
      : `${'  '.repeat(indentLevel)}<${transformedTag}${propertiesString}${eventListenersString} />`;

  const scripts = eventEntries.length > 0 ? generateReactControlledScript(tag, eventEntries, initialState) : null;
  const states = scripts ? [scripts.states, ...childStates] : childStates;
  const eventHandlers = scripts ? [scripts.eventHandler, ...childEventHandlers] : childEventHandlers;

  return { markup, states, eventHandlers, pdsComponents: [...pdsComponents, ...childPDSComponents] };
};

type ReactScripts = { states: string; eventHandler: string };

export const generateReactControlledScript = (
  tagName: string,
  eventEntries: [string, EventConfig][],
  initialState: StoryState<HTMLTagOrComponent>
): ReactScripts => {
  const states = eventEntries
    // Only create state if the current element's tagName is the same as the element the state is applied to e.g. don't create state for p-button onClick open flyout
    .filter(([_, { target }]) => tagName === target)
    .map(
      ([_, { prop }]) =>
        `  const [${prop}, set${pascalCase(prop)}] = useState(${(initialState?.properties as any)?.[prop]});`
    )
    .join('\n');

  const eventHandler = eventEntries
    .map(([eventName, { prop, value, eventValueKey, eventType, negateValue }]) => {
      if (eventValueKey) {
        return `  const ${eventName} = (e: ${eventType}) => {
    set${pascalCase(prop)}(${negateValue ? '!' : ''}e.detail.${eventValueKey});
  }`;
      }
      return `  const ${eventName} = () => {
    set${pascalCase(prop)}(${negateValue ? '!' : ''}${value});
  }`;
    })
    .join('\n');

  return { states, eventHandler };
};

export const generateReactProperties = (
  properties: HTMLElementOrComponentProps<HTMLTagOrComponent>,
  eventEntries: [string, EventConfig][]
) => {
  return Object.entries(properties)
    .map(([key, value]) => {
      if (eventEntries.some(([_, { prop }]) => prop === key)) {
        return ` ${key}={${key}}`;
      }
      if (typeof value === 'string') return ` ${key}="${value}"`;
      return ` ${key}={${JSON.stringify(value)}}`;
    })
    .join('');
};
