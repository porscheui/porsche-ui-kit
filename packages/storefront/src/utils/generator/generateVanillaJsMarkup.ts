import type {
  ElementConfig,
  EventConfig,
  EventsConfig,
  HTMLElementOrComponentProps,
  HTMLTagOrComponent,
} from '@/utils/generator/generator';
import { camelCase, kebabCase } from 'change-case';
import type { CSSProperties } from 'react';

const getVanillaJsCode = (code: string, selector?: string, eventHandlers?: string) => `<!doctype html>
<html lang="en">
<head>
  <title></title>
</head>
<body>

${code ?? ''}

<script>
${[selector, eventHandlers].filter(Boolean).join('\n')}
</script>
</body>
</html>`;

export const generateVanillaJsMarkup = (
  configs: (string | ElementConfig<HTMLTagOrComponent> | undefined)[],
  indentLevel = 0
): string => {
  const { markup, selector, eventHandlers } = configs.reduce(
    (acc, config) => {
      const result = createVanillaJSMarkup(config, indentLevel);
      acc.markup.push(result.markup);
      acc.selector.push(...result.selector);
      acc.eventHandlers.push(...result.eventHandlers);
      return acc;
    },
    { markup: [], selector: [], eventHandlers: [] } as { markup: string[]; selector: string[]; eventHandlers: string[] }
  );

  return getVanillaJsCode(markup.join('\n\n'), selector.join('\n'), eventHandlers.join('\n'));
};

const createVanillaJSMarkup = (
  config: string | ElementConfig<HTMLTagOrComponent> | undefined,
  indentLevel = 0
): { markup: string; selector: string[]; eventHandlers: string[] } => {
  if (!config) return { markup: '', selector: [], eventHandlers: [] };
  const indent = '  '.repeat(indentLevel);

  if (typeof config === 'string') return { markup: `${indent}${config}`, selector: [], eventHandlers: [] };

  const { tag, properties = {}, events = {}, children = [] } = config;

  const eventEntries = Object.entries(events);
  const propertyString = generateVanillaJsProperties(properties, eventEntries);

  const childrenMarkup = children.map((child) => createVanillaJSMarkup(child, indentLevel + 1));

  const markup =
    children.length > 0
      ? `${indent}<${tag}${propertyString}>\n${childrenMarkup.map(({ markup }) => markup).join('\n')}\n${indent}</${tag}>`
      : `${indent}<${tag}${propertyString} />`;

  const scripts = Object.keys(events).length > 0 ? generateVanillaJSControlledScript(tag, eventEntries) : null;

  return {
    markup,
    selector: scripts
      ? [scripts.selector, ...childrenMarkup.flatMap(({ selector }) => selector)]
      : childrenMarkup.flatMap(({ selector }) => selector),
    eventHandlers: scripts
      ? [scripts.eventHandler, ...childrenMarkup.flatMap(({ eventHandlers }) => eventHandlers)]
      : childrenMarkup.flatMap(({ eventHandlers }) => eventHandlers),
  };
};

export const generateVanillaJSControlledScript = (
  tagName: HTMLTagOrComponent,
  eventEntries: [string, EventConfig][]
) => {
  const constant = camelCase(tagName.replace('p-', ''));
  const selector = `  const ${constant} = document.querySelector("${tagName}");`;

  const eventHandler = eventEntries
    .map(([eventName, { target, prop, value, eventValueKey, negateValue }]) => {
      const element = camelCase(target.replace('p-', ''));
      const nativeEventName = camelCase(eventName.replace('on', ''));
      return eventValueKey
        ? `  ${constant}.addEventListener('${nativeEventName}', (e) => e.target.${prop} = ${negateValue ? '!' : ''}e.detail.${eventValueKey});`
        : `  ${constant}.addEventListener('${nativeEventName}', () => (${element}.${prop} = ${negateValue ? '!' : ''}${value}));`;
    })
    .join('\n');

  return { selector, eventHandler };
};

export const generateVanillaJsProperties = (
  properties: HTMLElementOrComponentProps<HTMLTagOrComponent>,
  eventEntries: [string, EventConfig][]
) => {
  return Object.entries(properties)
    .filter(([key]) => !eventEntries.some(([_, { prop }]) => prop === key))
    .map(([key, value]) => {
      if (typeof value === 'string') return ` ${kebabCase(key === 'className' ? 'class' : key)}="${value}"`;
      if (key === 'style')
        return ` style="${Object.entries(value as CSSProperties)
          .map(
            ([styleKey, styleValue]) => `${styleKey.startsWith('--') ? styleKey : kebabCase(styleKey)}: ${styleValue}`
          )
          .join('; ')}"`;
      if (key === 'aria') {
        return ` ${key}="${JSON.stringify(value).replace(/"/g, "'")}"`;
      }
      return ` ${kebabCase(key)}="${JSON.stringify(value)}"`;
    })
    .join('');
};
