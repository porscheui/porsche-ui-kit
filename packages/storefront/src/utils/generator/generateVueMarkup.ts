import type { StoryState } from '@/models/story';
import type {
  ConfiguratorTagNames,
  ElementConfig,
  EventConfig,
  EventsConfig,
  HTMLElementOrComponentProps,
  HTMLTagOrComponent,
} from '@/utils/generator/generator';
import { camelCase, kebabCase, pascalCase } from 'change-case';

const getVueCode = (
  imports: string,
  states: string | undefined,
  eventHandlers: string | undefined,
  code: string | undefined
) => `<script setup lang="ts">\n${imports}\n${states ? `\n${states}\n` : ''}${eventHandlers ? `\n${eventHandlers}\n` : ''}</script>

<template>
${code}
</template>`;

export const generateVueMarkup = (
  configs: (string | ElementConfig<HTMLTagOrComponent> | undefined)[],
  initialState: StoryState<HTMLTagOrComponent>,
  indentLevel = 1
): string => {
  const results = configs.map((config) => createVueMarkup(config, initialState, indentLevel));
  const markup = results.map(({ markup }) => markup).join('\n\n');
  const states = results
    .flatMap(({ states }) => states)
    .filter((state) => state)
    .join('\n');
  const eventHandlers = results.flatMap(({ eventHandlers }) => eventHandlers).join('\n');
  const pdsComponents = new Set(results.flatMap(({ pdsComponents }) => pdsComponents));
  const imports = `${states.length > 0 ? "  import { ref } from 'vue';\n" : ''}${
    pdsComponents.size > 0
      ? `  import { ${Array.from(pdsComponents).sort().join(', ')} } from '@porsche-design-system/components-vue';`
      : ''
  }`;

  return getVueCode(imports, states, eventHandlers, markup);
};

const createVueMarkup = (
  config: string | ElementConfig<HTMLTagOrComponent> | undefined,
  initialState: StoryState<HTMLTagOrComponent>,
  indentLevel = 0
): { markup: string; states: string[]; eventHandlers: string[]; pdsComponents: string[] } => {
  if (!config) return { markup: '', states: [], eventHandlers: [], pdsComponents: [] };
  const indent = '  '.repeat(indentLevel);

  if (typeof config === 'string')
    return { markup: `${indent}${config}`, states: [], eventHandlers: [], pdsComponents: [] };

  const { tag, properties = {}, events = {}, children = [] } = config;
  const isPDSComponent = tag.startsWith('p-');

  let transformedTag: string = tag;
  const pdsComponents: string[] = [];
  if (isPDSComponent) {
    transformedTag = pascalCase(tag);
    pdsComponents.push(transformedTag);
  }

  const eventEntries = Object.entries(events);
  const propertiesString = generateVueProperties(properties, eventEntries);

  const eventListenersString =
    eventEntries.length > 0
      ? ` ${eventEntries.map(([eventName]) => `@${camelCase(eventName.replace('on', ''))}="${eventName}"`).join(' ')}`
      : '';

  const childrenResults = children.map((child) => createVueMarkup(child, initialState, indentLevel + 1));
  const childMarkup = childrenResults.map(({ markup }) => markup).join('\n');
  const childStates = childrenResults.flatMap(({ states }) => states);
  const childEventHandlers = childrenResults.flatMap(({ eventHandlers }) => eventHandlers);
  const childPDSComponents = childrenResults.flatMap(({ pdsComponents }) => pdsComponents);

  const markup =
    children.length > 0
      ? `${'  '.repeat(indentLevel)}<${transformedTag}${propertiesString}${eventListenersString}>\n${childMarkup}\n${'  '.repeat(indentLevel)}</${transformedTag}>`
      : `${'  '.repeat(indentLevel)}<${transformedTag}${propertiesString}${eventListenersString} />`;

  const scripts = eventEntries.length > 0 ? generateVueControlledScript(tag, eventEntries, initialState) : null;
  const states = scripts ? [scripts.states, ...childStates] : childStates;
  const eventHandlers = scripts ? [scripts.eventHandler, ...childEventHandlers] : childEventHandlers;

  return { markup, states, eventHandlers, pdsComponents: [...pdsComponents, ...childPDSComponents] };
};

type VueScripts = { states: string; eventHandler: string };

export const generateVueControlledScript = (
  tagName: string,
  eventEntries: [string, EventConfig][],
  initialState: StoryState<HTMLTagOrComponent>
): VueScripts => {
  const states = eventEntries
    // Only create state if the current element's tagName is the same as the element the state is applied to e.g. don't create state for p-button onClick open flyout
    .filter(([_, { target }]) => tagName === target)
    .map(([_, { prop }]) => `  const ${prop} = ref(${(initialState?.properties as any)?.[prop]});`)
    .join('\n');

  const eventHandler = eventEntries
    .map(([eventName, { prop, value, eventValueKey, eventType, negateValue }]) => {
      if (eventValueKey) {
        return `  const ${eventName} = (e: ${eventType}) => {
    ${prop}.value = ${negateValue ? '!' : ''}e.detail.${eventValueKey};
  }`;
      }
      return `  const ${eventName} = () => {
    ${prop}.value = ${negateValue ? '!' : ''}${value};
  }`;
    })
    .join('\n');

  return { states, eventHandler };
};

export const generateVueProperties = (
  properties: HTMLElementOrComponentProps<HTMLTagOrComponent>,
  eventEntries: [string, EventConfig][]
) => {
  return Object.entries(properties)
    .map(([key, value]) => {
      if (eventEntries.some(([_, { prop }]) => prop === key)) {
        return ` :${key}="${key}"`;
      }
      if (typeof value === 'string') return ` ${key}="${value}"`;
      if (key === 'aria') {
        return ` ${key}="${JSON.stringify(value).replace(/"/g, "'")}"`;
      }
      return ` :${key}="${JSON.stringify(value)}"`;
    })
    .join('');
};
