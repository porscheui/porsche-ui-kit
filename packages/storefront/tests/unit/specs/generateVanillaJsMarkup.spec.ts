import type { CSSProperties } from 'react';
import {
  generateVanillaJSControlledScript,
  generateVanillaJsMarkup,
  generateVanillaJsProperties,
} from '../../../src/utils/generator/generateVanillaJsMarkup';
import { buttonTestConfig, flyoutTestConfig } from '../data/generator.testdata';

describe('generateVanillaJsMarkup()', () => {
  it('should generate correct Vanilla JS markup for button', () => {
    const output = generateVanillaJsMarkup(buttonTestConfig);
    expect(output).toMatchSnapshot();
  });
  it('should generate correct Vanilla JS markup for flyout', () => {
    const output = generateVanillaJsMarkup(flyoutTestConfig);
    expect(output).toMatchSnapshot();
  });
});

describe('generateVanillaJSControlledScript()', () => {
  it('should return correct selector & eventHandler for direct value', () => {
    const { selector, eventHandler } = generateVanillaJSControlledScript('p-flyout', [
      [
        'onDismiss',
        {
          target: 'p-flyout',
          prop: 'open',
          value: false,
        },
      ],
    ]);
    expect(selector).toMatchInlineSnapshot('"  const flyout = document.querySelector("p-flyout");"');
    expect(eventHandler).toMatchInlineSnapshot(
      '"  flyout.addEventListener(\'dismiss\', () => (flyout.open = false));"'
    );
  });
  it('should return correct selector & eventHandler for event value', () => {
    const { selector, eventHandler } = generateVanillaJSControlledScript('p-accordion', [
      [
        'onUpdate',
        {
          target: 'p-accordion',
          prop: 'open',
          eventValueKey: 'open',
          eventType: 'CustomEvent<AccordionUpdateEventDetail>',
        },
      ],
    ]);
    expect(selector).toMatchInlineSnapshot('"  const accordion = document.querySelector("p-accordion");"');
    expect(eventHandler).toMatchInlineSnapshot(
      '"  accordion.addEventListener(\'update\', (e) => e.target.open = e.detail.open);"'
    );
  });
  it('should return correct selector & eventHandler for event value with negated value', () => {
    const { selector, eventHandler } = generateVanillaJSControlledScript('p-link-tile-product', [
      [
        'onLike',
        {
          target: 'p-link-tile-product',
          prop: 'liked',
          eventValueKey: 'liked',
          eventType: 'CustomEvent<LinkTileProductLikeEvent>',
          negateValue: true,
        },
      ],
    ]);
    expect(selector).toMatchInlineSnapshot(
      '"  const linkTileProduct = document.querySelector("p-link-tile-product");"'
    );
    expect(eventHandler).toMatchInlineSnapshot(
      '"  linkTileProduct.addEventListener(\'like\', (e) => e.target.liked = !e.detail.liked);"'
    );
  });

  it('should return correct selector & eventHandler when target is not current tagName', () => {
    const { selector, eventHandler } = generateVanillaJSControlledScript('p-button', [
      [
        'onClick',
        {
          target: 'p-flyout',
          prop: 'open',
          value: true,
        },
      ],
    ]);
    expect(selector).toMatchInlineSnapshot('"  const button = document.querySelector("p-button");"');
    expect(eventHandler).toMatchInlineSnapshot('"  button.addEventListener(\'click\', () => (flyout.open = true));"');
  });
});

describe('generateVanillaJsProperties()', () => {
  it('should generate correct properties', () => {
    const props = generateVanillaJsProperties(
      {
        aria: { 'aria-haspopup': true, 'aria-label': 'Some more descriptive label' },
        type: 'button',
        compact: true,
        icon: 'add',
      },
      []
    );
    expect(props).toMatchInlineSnapshot(
      '" aria="{\'aria-haspopup\':true,\'aria-label\':\'Some more descriptive label\'}" type="button" compact="true" icon="add""'
    );
  });

  it('should generate correct properties for style prop', () => {
    const props = generateVanillaJsProperties(
      {
        style: { backgroundColor: 'red', '--custom-prop': '1px' } as CSSProperties,
      },
      []
    );
    expect(props).toMatchInlineSnapshot('" style="background-color: red; --custom-prop: 1px""');
  });

  it('should remove props included in events', () => {
    const props = generateVanillaJsProperties(
      {
        open: true,
        name: 'Some prop',
      },
      [
        [
          'onClick',
          {
            target: 'p-flyout',
            prop: 'open',
          },
        ],
      ]
    );
    expect(props).toMatchInlineSnapshot('" name="Some prop""');
  });
});
