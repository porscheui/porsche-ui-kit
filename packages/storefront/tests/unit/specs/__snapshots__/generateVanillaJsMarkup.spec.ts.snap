// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateVanillaJsMarkup() > should generate correct Vanilla JS markup for button 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <title></title>
</head>
<body>

<p-button aria="{'aria-haspopup':true,'aria-label':'Some more descriptive label'}" type="button" compact="true" icon="add">
  Open Banner
</p-button>

<p-banner dismiss-button="false" heading="Some heading" heading-tag="h4" state="warning" />

<script>
  const button = document.querySelector("p-button");
  const banner = document.querySelector("p-banner");
  button.addEventListener('click', () => (banner.open = true));
  banner.addEventListener('dismiss', () => (banner.open = false));
</script>
</body>
</html>"
`;

exports[`generateVanillaJsMarkup() > should generate correct Vanilla JS markup for flyout 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <title></title>
</head>
<body>

<p-button type="button" aria="{'aria-haspopup':'dialog'}">
  Open Flyout
</p-button>

<p-flyout aria="{'aria-label':'Some Heading'}">
  <p-heading slot="header" size="large" tag="h2">
    Some Heading
  </p-heading>
  <p-text>
    Some Content Begin
  </p-text>
  <div style="width: 10px; height: 120vh; background: deeppink" />
  <p-text>
    Some Content End
  </p-text>
  <p-button-group slot="footer">
    <p-button type="button">
      Proceed
    </p-button>
    <p-button type="button" variant="secondary">
      Cancel
    </p-button>
  </p-button-group>
  <p-text slot="sub-footer">
    Some additional Sub-Footer
  </p-text>
</p-flyout>

<script>
  const button = document.querySelector("p-button");
  const flyout = document.querySelector("p-flyout");
  button.addEventListener('click', () => (flyout.open = true));
  flyout.addEventListener('dismiss', () => (flyout.open = false));
</script>
</body>
</html>"
`;
